Enhancement: Option to change OpenAI API Endpoint

hahayusuf opened the pull request:
You can now set a custom API base URL through the env file using the `API_BASE_URL` variable. When kept blank (""), it will use the default base URL as per the previous implementation.

The README, and setup scripts were updated to incorporate the new variable.

pengmaochang commented:
@hahayusuf the file "setup.sh" write like this may be better
```
#!/bin/bash
cd "$(dirname "$0")" || exit

echo -n "Enter your OpenAI Key (eg: sk...): "
read OPENAI_API_KEY
echo -n "Enter your API_BASE_URL IS YOU NEEDED ( press enter for default): "
read API_BASE_URL

NEXTAUTH_SECRET=$(openssl rand -base64 32)

ENV="NODE_ENV=development\n\
NEXTAUTH_SECRET=$NEXTAUTH_SECRET\n\
NEXTAUTH_URL=http://localhost:3000\n\
OPENAI_API_KEY=$OPENAI_API_KEY\n\
DATABASE_URL=file:../db/db.sqlite\n
API_BASE_URL=$API_BASE_URL\n"
...
```

awtkns commented on src/env/schema.mjs:
```diff
@@ -20,7 +20,8 @@ export const serverSchema = z.object({
   ),
-  OPENAI_API_KEY: z.string()
+  OPENAI_API_KEY: z.string(),
+  API_BASE_URL: z.string()
```
Might be best to prefix this with OPENAI_ as we are doing in the field above :)

hahayusuf replied:
I have resolved this with the last commit.

Mohamedhoussein commented:
I am interested

Jshen123 commented on src/components/ChatWindow.tsx:
```diff
@@ -123,16 +124,29 @@ const MacWindowHeader = () => {
       return;
     }
 
+    element.innerText="Copied!"
```
Hi @hahayusuf ,  change this to "text" param. This is hardcoded right now.

Jshen123 commented on src/components/ChatWindow.tsx:
```diff
@@ -161,10 +175,10 @@ const MacWindowHeader = () => {
       >
         <FaClipboard size={12} />
-        <p className="font-mono">Copy</p>
+        <p id="btn-copy-01" className="font-mono">Copy</p>
```
following the [BEM](https://getbem.com/naming/) convention, can you change this to something like "copy--button"

Jshen123 commented on src/utils/chain.ts:
```diff
@@ -2,14 +2,31 @@ import { OpenAI } from "langchain/llms/openai";
+    basePath : baseUrl
+  })
+
+};
```
Hi @hahayusuf, you can probably refactor this to something like this:

   ```
 return new OpenAI({
      openAIApiKey:
        customApiKey === "" ? process.env.OPENAI_API_KEY : customApiKey,
      temperature: 0.9,
      modelName: "gpt-3.5-turbo",
      maxTokens: 300,
    }, {
    basePath : process.env.OPENAI_API_BASE_URL || undefined
  })
```

Jshen123 reviewed (changes requested):
Hi @hahayusuf, thanks for making this PR :) this is a very useful change

I have included some feedback. Also, there will likely be merge conflicts since a lot has changed since this PR was made

jondwillis commented:
This is also useful for analytics and caching layers such as https://helicone.ai

hahayusuf commented:
> Hi @hahayusuf, thanks for making this PR :) this is a very useful change
> 
> 
> 
> I have included some feedback. Also, there will likely be merge conflicts since a lot has changed since this PR was made


Thanks for your feedback! 

Will let you know when it's ready to merge.

hahayusuf commented:
@Jshen123  

The merge conflicts have now been resolved and can be merged automatically.

I have refactored this to use the `ModelSettings` interface for setting the `customBaseUrl`.

- Input field created for customBaseUrl under advanced settings in settings dialog.
- Fixed other comments related to "Copy!" message when clicking the copy button.

Jshen123 reviewed (changes requested):
Hi @hahayusuf  loooks really good! 🎉  I made some minor tweaks here. (you can see the changes once you pull into your local branch)
https://github.com/reworkd/AgentGPT/pull/246

I highlighted the main ones. everything else is just automatic syntax formatting from linter

hahayusuf commented:
@Jshen123 

Thanks for your feedback. 

I have added your "minor tweaks" commit. 

Then I noticed some merge conflicts, which I have now resolved.

Jshen123 reviewed (approved):
approved!

awtkns commented:
I will take a look at this shortly!

awtkns commented on src/utils/prompts.ts:
```diff
@@ -13,6 +13,12 @@ export const createModel = (settings: ModelSettings) =>
+  },
+  {
+    basePath:
+      settings.customBaseUrl === ""
+        ? process.env.OPENAI_API_BASE_URL
```
```
env. OPENAI_API_BASE_URL
```

awtkns replied:
```suggestion
      settings.customBaseUrl === ""
        ? env.OPENAI_API_BASE_URL
```
Above we will need to import the client schema from `env/client.mjs`. Doing so allows us to preparse and validate the environment variables. More info can be found here: https://create.t3.gg/en/usage/env-variables

awtkns reviewed (changes requested):
I have a slight security concern about allowing users to change the base url in our production environments. Doing so could potentially expose our keys if we allow them to change the baseurl + not have to input their own key. Ideally we will probably want to keep this client side if possible. Thoughts @asim-shrestha?

hahayusuf commented:
> I have a slight security concern about allowing users to change the base url in our production environments. Doing so could potentially expose our keys if we allow them to change the baseurl + not have to input their own key. Ideally we will probably want to keep this client side if possible. Thoughts @asim-shrestha?

It is a valid concern @awtkns. We can protect against it by implementing two safeguards:
-Frontend: hiding the custom url input unless a `customApiKey` is set.
-Backend: when `settings.customBaseUrl` is set, `env.OPENAI_API_KEY` cannot be used, only `customApiKey`. A simple guard clause should do the trick.

Jshen123 commented:
> > I have a slight security concern about allowing users to change the base url in our production environments. Doing so could potentially expose our keys if we allow them to change the baseurl + not have to input their own key. Ideally we will probably want to keep this client side if possible. Thoughts @asim-shrestha?
> 
> It is a valid concern @awtkns. We can protect against it by implementing two safeguards: -Frontend: hiding the custom url input unless a `customApiKey` is set. -Backend: when `settings.customBaseUrl` is set, `env.OPENAI_API_KEY` cannot be used, only `customApiKey`. A simple guard clause should do the trick.

@awtkns that's a really good callout. @hahayusuf  agree with your suggested solution (only change here would be to use either disable or disable + hide)

@awtkns @asim-shrestha, if we are not comfortable with custom URL in production, perhaps we can rollback URL input feature and release the rest of the code (so users can at least use their own URLs locally). Thoughts?

hahayusuf commented:
Update: I have implemented both safeguards discussed. 

Thanks for your feedback @Jshen123 

Look forward to hearing what others think.

asim-shrestha commented:
Hey sorry @hahayusuf, this has flown under the radar. Unfortunate because it seems we have a bunch more conflicts. I can have a look into manually fixing this stuff if you're not able. 

Configuring endpoint locally via ENV seems like a good win. For endpoint in settings, given the safe guards it should be ok but maybe some unit tests are in order to validate this. Otherwise, perhaps we can provide a dropdown of options people can use. Not sure what other endpoints they might try.

hahayusuf commented:
Hey @asim-shrestha 

Thanks for offering to resolve some of the conflicts.

I agree, using env to set a custom base url should be a good enough win.

Will work on that. May need help ironing out conflicts.

asim-shrestha commented:
Awesome!

awtkns commented:
Closed as this has been already implemented
